Notes
====

### Tasks:

1. Modify ```get_codon_usage_frequencies()``` to store table data for the entire chromosome 

2. Write a function that creates a new protein sequence from the coding sequence if the protein is not aligned

    - You may need to create a dummy database (just in python) by using regex on the chromosome file, that just contains the coordinates, dna and translated protein



### Follow up notes

Should get_entries() for a given gene id (or other search) only ever have one genbank entry in the result? i.e. the returned list only has one dictionary?
